File: hs/Protocol.hs
-- | Protocol implements a subset of the memcache binary protocol
module Protocol(
    -- * Types
    Packet,
    -- * Functions
    readPacket, writeRequest, writeResponse,
    packetCmd, packetStatus, packetOpaque, packetVersion,
    packetExtras, packetKey, packetValue,
    makeExpiry, makeFlags, getExpiry, getFlags,
    -- * Constants
    Magic, magicRequest, magicResponse,
    Cmd, cmdGet, cmdSet, cmdAdd, cmdReplace, cmdDelete, cmdQuit,
    Status, statusNoError, statusKeyNotFound, statusKeyExists,
    statusItemNotStored, statusNotSupported, statusInternalError
    ) where

import Data.ByteString(ByteString,index,pack)
import qualified Data.ByteString as B
import Data.Maybe(catMaybes)
import Data.Time(NominalDiffTime)
import Data.Word(Word8,Word16,Word32,Word64)
import Network.Socket(Socket)
import Network.Socket.ByteString(recv,sendMany)

import Logger(Logger)

-- | Packet
data Packet = Packet ByteString [ByteString] (Maybe ByteString) (Maybe ByteString)

-- | readPacket
readPacket :: Logger -> Socket -> Magic -> IO (Maybe Packet)
readPacket logger socket (Magic magic) = do
    maybeHeader <- readChunk logger socket 24
    maybe (return Nothing) readExtras maybeHeader
  where
    readExtras header
      | header `index` 0 /= magic = return Nothing
      | extrasLength header > 0 = do
            maybeExtras <- readChunk logger socket (extrasLength header)
            maybe (return Nothing) (readKey header . Packet header . (:[])) maybeExtras
      | otherwise = readKey header (Packet header [])
    readKey header packet
      | keyLength header > 0 = do
            maybeKey <- readChunk logger socket (keyLength header)
            maybe (return Nothing) (readValue header . packet . Just) maybeKey
      | otherwise = readValue header (packet Nothing)
    readValue header packet
      | valueLength header > 0 = do
            maybeValue <- readChunk logger socket (valueLength header)
            maybe (return Nothing) (return . Just . packet . Just) maybeValue
      | otherwise = return (Just (packet Nothing))
    extrasLength header = fromIntegral (header `index` 4)
    keyLength header = fromIntegral (header `readWord16` 2)
    totalLength header = fromIntegral (header `readWord32` 8)
    valueLength header = totalLength header - extrasLength header - keyLength header

readChunk :: Logger -> Socket -> Int -> IO (Maybe ByteString)
readChunk logger socket size = readChunk' size []
  where
    readChunk' n list = do
        bytes <- recv socket n
        logger "<" bytes
        if B.length bytes == 0
          then return Nothing
          else
            if B.length bytes < n
              then readChunk' (n - B.length bytes) (bytes:list)
              else
                if null list
                  then return $ Just bytes
                  else return $ Just $ B.concat $ reverse (bytes:list)

-- | writeRequest
writeRequest :: Logger -> Socket -> Cmd -> Word32 -> Word64 -> [ByteString] -> Maybe ByteString -> Maybe ByteString -> IO ()
writeRequest logger socket (Cmd cmd) opaque version extras key value = do
    let bytes = (pack [magic, cmd,
                       keyLength `shift` 1, keyLength `shift` 0,
                       extrasLength `shift` 0, 0,
                       0, 0,
                       totalLength `shift` 3, totalLength `shift` 2,
                       totalLength `shift` 1, totalLength `shift` 0,
                       opaque `shift` 3, opaque `shift` 2,
                       opaque `shift` 1, opaque `shift` 0,
                       version `shift` 7, version `shift` 6,
                       version `shift` 5, version `shift` 4,
                       version `shift` 3, version `shift` 2,
                       version `shift` 1, version `shift` 0]
        	 : extras ++ catMaybes [key,value])
    mapM_ (logger ">") bytes
    sendMany socket bytes
  where
    Magic magic = magicRequest
    keyLength = maybe 0 B.length key
    extrasLength = sum $ map B.length extras
    totalLength = keyLength + extrasLength + maybe 0 B.length value

-- | writeResponse
writeResponse :: Logger -> Socket -> Packet -> Status -> Word64 -> [ByteString] -> Maybe ByteString -> Maybe ByteString -> IO ()
writeResponse logger socket (Packet header _ _ _) (Status status) version extras key value = do
    let bytes = (pack [magic, header `index` 1,
                       keyLength `shift` 1, keyLength `shift` 0,
                       extrasLength `shift` 0, 0,
                       status `shift` 1, status `shift` 0,
                       totalLength `shift` 3, totalLength `shift` 2,
                       totalLength `shift` 1, totalLength `shift` 0,
                       header `index` 12, header `index` 13,
                       header `index` 14, header `index` 15,
                       version `shift` 7, version `shift` 6,
                       version `shift` 5, version `shift` 4,
                       version `shift` 3, version `shift` 2,
                       version `shift` 1, version `shift` 0]
        	 : extras ++ catMaybes [key,value])
    mapM_ (logger ">") bytes
    sendMany socket bytes
  where
    Magic magic = magicResponse
    keyLength = maybe 0 B.length key
    extrasLength = sum $ map B.length extras
    totalLength = keyLength + extrasLength + maybe 0 B.length value

-- | packetCmd
packetCmd :: Packet -> Cmd
packetCmd (Packet header _ _ _) = Cmd $ header `index` 1

-- | packetStatus
packetStatus :: Packet -> Status
packetStatus (Packet header _ _ _) = Status $ header `readWord16` 6

-- | packetOpaque
packetOpaque :: Packet -> Word32
packetOpaque (Packet header _ _ _) = header `readWord32` 12

-- | packetVersion
packetVersion :: Packet -> Word64
packetVersion (Packet header _ _ _) = header `readWord64` 16

-- | packetExtras
packetExtras :: Packet -> [ByteString]
packetExtras (Packet _ extras _ _) = extras

-- | packetKey
packetKey :: Packet -> Maybe ByteString
packetKey (Packet _ _ key _) = key

-- | packetValue
packetValue :: Packet -> Maybe ByteString
packetValue (Packet _ _ _ value) = value

-- | makeExpiry
makeExpiry :: NominalDiffTime -> ByteString
makeExpiry diffTime =
    pack [dt `shift` 3, dt `shift` 2, dt `shift` 1, dt `shift` 0]
  where
    dt = floor diffTime

-- | makeFlags
makeFlags :: Word32 -> ByteString
makeFlags f = pack [f `shift` 3, f `shift` 2, f `shift` 1, f `shift` 0]

-- | getExpiry
getExpiry :: ByteString -> Int -> NominalDiffTime
getExpiry extras index = fromInteger $ fromIntegral $ extras `readWord32` index

-- | getFlags
getFlags :: ByteString -> Int -> Word32
getFlags extras index = extras `readWord32` index

shift :: (Integral i,Integral r) => i -> i -> r
shift a b = fromIntegral $ a `div` 256^b

readWord16 :: ByteString -> Int -> Word16
readWord16 bytes i =
    256*fromIntegral (bytes `index` i) + fromIntegral (bytes `index` (i + 1))

readWord32 :: ByteString -> Int -> Word32
readWord32 bytes i =
          256^3*fromIntegral (bytes `index` i)
        + 256^2*fromIntegral (bytes `index` (i + 1))
        + 256  *fromIntegral (bytes `index` (i + 2))
        +       fromIntegral (bytes `index` (i + 3))

readWord64 :: ByteString -> Int -> Word64
readWord64 bytes i =
          256^7*fromIntegral (bytes `index` i)
        + 256^6*fromIntegral (bytes `index` (i + 1))
        + 256^5*fromIntegral (bytes `index` (i + 2))
        + 256^4*fromIntegral (bytes `index` (i + 3))
        + 256^3*fromIntegral (bytes `index` (i + 4))
        + 256^2*fromIntegral (bytes `index` (i + 5))
        + 256  *fromIntegral (bytes `index` (i + 6))
        +       fromIntegral (bytes `index` (i + 7))

newtype Magic = Magic Word8

magicRequest :: Magic
magicRequest = Magic 0x80

magicResponse :: Magic
magicResponse = Magic 0x81

newtype Cmd = Cmd Word8 deriving Eq

cmdGet :: Cmd
cmdGet = Cmd 0

cmdSet :: Cmd
cmdSet = Cmd 1

cmdAdd :: Cmd
cmdAdd = Cmd 2

cmdReplace :: Cmd
cmdReplace = Cmd 3

cmdDelete :: Cmd
cmdDelete = Cmd 4

cmdQuit :: Cmd
cmdQuit = Cmd 7

newtype Status = Status Word16 deriving Eq

statusNoError :: Status
statusNoError = Status 0

statusKeyNotFound :: Status
statusKeyNotFound = Status 1

statusKeyExists :: Status
statusKeyExists = Status 2

statusItemNotStored :: Status
statusItemNotStored = Status 5

statusNotSupported :: Status
statusNotSupported = Status 131

statusInternalError :: Status
statusInternalError = Status 132

File: go/protocol/protocol.go
// Package protocol supports a subset of the memcache protocol.
package protocol

import (
	"errors"
	"io"
	"log"
	"time"
)

const (
	// Magic byte values.
	MagicRequest  = 0x80
	MagicResponse = 0x81

	// Response status values.
	StatusNoError       = 0
	StatusKeyNotFound   = 1
	StatusKeyExists     = 2
	StatusItemNotStored = 5
	StatusNotSupported  = 131
	StatusInternalError = 132

	// Command opcode byte values.
	CmdGet     = 0
	CmdSet     = 1
	CmdAdd     = 2
	CmdReplace = 3
	CmdDelete  = 4
	CmdQuit    = 7
)

var BadMagic = errors.New("Bad magic")

// Packet represents a packet in the memcache protocol.
type Packet struct {
	header             []byte
	Extras, Key, Value []byte
}

// ReadPacket reads a packet in the memcache protocol.  magic is the
// magic byte to expect.  If header is large enough to contain the
// packet header (24 bytes), it will be used to hold the header.
func ReadPacket(logger *log.Logger, reader io.Reader, magic byte, header []byte) (*Packet, error) {
	if header == nil || len(header) < 24 {
		header = make([]byte, 24)
	} else {
		header = header[:24]
	}
	err := readChunk(logger, reader, header)
	if err != nil {
		return nil, err
	}
	if header[0] != magic {
		return nil, BadMagic
	}
	packet := &Packet{header: header}
	extrasLen := int(header[4])
	keyLen := int(getUint16(header, 2))
	totalLen := int(getUint32(header, 8))
	valueLen := totalLen - extrasLen - keyLen
	if extrasLen > 0 {
		packet.Extras = make([]byte, extrasLen)
		err = readChunk(logger, reader, packet.Extras)
		if err != nil {
			return nil, err
		}
	}
	if keyLen > 0 {
		packet.Key = make([]byte, keyLen)
		err = readChunk(logger, reader, packet.Key)
		if err != nil {
			return nil, err
		}
	}
	if valueLen > 0 {
		packet.Value = make([]byte, valueLen)
		err = readChunk(logger, reader, packet.Value)
		if err != nil {
			return nil, err
		}
	}
	return packet, nil
}

// WriteRequest writes a request packet in the memcache protocol.
func WriteRequest(logger *log.Logger, writer io.Writer, cmd byte, opaque uint32, version uint64, header, extras, key, value []byte) error {
	if header == nil || len(header) < 24 {
		header = make([]byte, 24)
	} else {
		header = header[:24]
	}
	header[0] = MagicRequest
	header[1] = cmd
	setUint16(header, 2, uint16(len(key)))
	header[4] = uint8(len(extras))
	header[5] = 0
	header[6] = 0
	header[7] = 0
	setUint32(header, 8, uint32(len(key)+len(extras)+len(value)))
	setUint32(header, 12, opaque)
	setUint64(header, 16, version)
	logChunk(logger, ">", header)
	_, err := writer.Write(header)
	if err != nil {
		return err
	}
	logChunk(logger, ">", extras)
	_, err = writer.Write(extras)
	if err != nil {
		return err
	}
	logChunk(logger, ">", key)
	_, err = writer.Write(key)
	if err != nil {
		return err
	}
	logChunk(logger, ">", value)
	_, err = writer.Write(value)
	return err
}

// WriteResponse writes a response packet in the memcache protocol.
// packet is the request packet.
func WriteResponse(logger *log.Logger, writer io.Writer, packet *Packet, status uint16, version uint64, extras, key, value []byte) error {
	packet.header[0] = MagicResponse
	setUint16(packet.header, 2, uint16(len(key)))
	packet.header[4] = uint8(len(extras))
	packet.header[5] = 0
	setUint16(packet.header, 6, status)
	setUint32(packet.header, 8, uint32(len(key)+len(extras)+len(value)))
	setUint64(packet.header, 16, version)
	logChunk(logger, ">", packet.header)
	_, err := writer.Write(packet.header)
	if err != nil {
		return err
	}
	logChunk(logger, ">", extras)
	_, err = writer.Write(extras)
	if err != nil {
		return err
	}
	logChunk(logger, ">", key)
	_, err = writer.Write(key)
	if err != nil {
		return err
	}
	logChunk(logger, ">", value)
	_, err = writer.Write(value)
	return err
}

func readChunk(logger *log.Logger, reader io.Reader, bytes []byte) error {
	for {
		n, err := reader.Read(bytes)
		if err != nil {
			return err
		}
		logChunk(logger, "<", bytes[:n])
		if n >= len(bytes) {
			return nil
		}
		bytes = bytes[n:]
	}
	return nil
}

func logChunk(logger *log.Logger, prefix string, bytes []byte) {
	if logger == nil {
		return
	}
	for i := 0; i < len(bytes); i += 16 {
		switch (len(bytes) - i) / 4 {
		case 0:
			logger.Printf("%s %x", prefix, bytes[i:])
		case 1:
			logger.Printf("%s %x %x", prefix, bytes[i:i+4], bytes[i+4:])
		case 2:
			logger.Printf("%s %x %x %x", prefix, bytes[i:i+4], bytes[i+4:i+8], bytes[i+8:])
		case 3:
			logger.Printf("%s %x %x %x %x", prefix, bytes[i:i+4], bytes[i+4:i+8], bytes[i+8:i+12], bytes[i+12:])
		default:
			logger.Printf("%s %x %x %x %x", prefix, bytes[i:i+4], bytes[i+4:i+8], bytes[i+8:i+12], bytes[i+12:i+16])
		}
	}
}

func getUint16(bytes []byte, index int) uint16 {
	return uint16(bytes[index])<<8 + uint16(bytes[index+1])
}

func getUint32(bytes []byte, index int) uint32 {
	return uint32(bytes[index])<<24 + uint32(bytes[index+1])<<16 + uint32(bytes[index+2])<<8 + uint32(bytes[index+3])
}

func getUint64(bytes []byte, index int) uint64 {
	return uint64(bytes[index])<<56 + uint64(bytes[index+1])<<48 + uint64(bytes[index+2])<<40 + uint64(bytes[index+3])<<32 + uint64(bytes[index+4])<<24 + uint64(bytes[index+5])<<16 + uint64(bytes[index+6])<<8 + uint64(bytes[index+7])
}

func setUint16(bytes []byte, index int, value uint16) {
	bytes[index] = byte(value >> 8)
	bytes[index+1] = byte(value)
}

func setUint32(bytes []byte, index int, value uint32) {
	bytes[index] = byte(value >> 24)
	bytes[index+1] = byte(value >> 16)
	bytes[index+2] = byte(value >> 8)
	bytes[index+3] = byte(value)
}

func setUint64(bytes []byte, index int, value uint64) {
	bytes[index] = byte(value >> 56)
	bytes[index+1] = byte(value >> 48)
	bytes[index+2] = byte(value >> 40)
	bytes[index+3] = byte(value >> 32)
	bytes[index+4] = byte(value >> 24)
	bytes[index+5] = byte(value >> 16)
	bytes[index+6] = byte(value >> 8)
	bytes[index+7] = byte(value)
}

// Cmd returns the command opcode byte.
func (packet *Packet) Cmd() byte {
	return packet.header[1]
}

// Status returns the response status.
func (packet *Packet) Status() uint16 {
	return getUint16(packet.header, 6)
}

// Opaque returns the opaque value.
func (packet *Packet) Opaque() uint32 {
	return getUint32(packet.header, 12)
}

// Version returns the data version check or CAS.
func (packet *Packet) Version() uint64 {
	return getUint64(packet.header, 16)
}

// Flags returns the flags, assuming the packet includes the flags.
func (packet *Packet) Flags() uint32 {
	return getUint32(packet.Extras, 0)
}

// Expiry returns the time to expiration in nanoseconds, assuming the
// packet includes the expiry.
func (packet *Packet) Expiry() uint64 {
	if len(packet.Extras) == 4 {
		return uint64(getUint32(packet.Extras, 0)) * uint64(time.Second)
	}
	return uint64(getUint32(packet.Extras, 4)) * uint64(time.Second)
}

// MakeExpiry makes the extras bytes containing the time to expiration.
func MakeExpiry(expiry, t uint64) []byte {
	extras := make([]byte, 4)
	setUint32(extras, 0, uint32((expiry-t)/uint64(time.Second)))
	return extras
}

// MakeFlags makes the extras bytes containing the flags.
func MakeFlags(flags uint32) []byte {
	extras := make([]byte, 4)
	setUint32(extras, 0, flags)
	return extras
}

// MakeFlagsExpiry makes the extras bytes containing the flags and the
// time to expiration.
func MakeFlagsExpiry(flags uint32, expiry, t uint64) []byte {
	extras := make([]byte, 8)
	setUint32(extras, 0, flags)
	setUint32(extras, 4, uint32((expiry-t)/uint64(time.Second)))
	return extras
}
